app/services/optout/OptOutService.scala

```diff
@@ -97,41 +92,62 @@ class OptOutService @Inject()(itsaStatusUpdateConnector: ITSAStatusUpdateConnect
     } yield optOutChecks
   }
 
-  def makeOptOutUpdateRequest(taxPayerIntent: Option[TaxYear] = None)
-                             (implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[OptOutUpdateResponse] = {
-
-    setupOptOutProposition().flatMap { proposition =>
-      if (proposition.isOneYearOptOut)
-        makeOptOutUpdateRequestForOneYear(proposition)
-
-      else {
-        taxPayerIntent.flatMap(intent => proposition.optOutTaxYearFor(intent))
-          .map(optOutTaxYearIntent => makeOptOutUpdateRequest(proposition, optOutTaxYearIntent))
-          .getOrElse(Future.successful(OptOutUpdateResponseFailure.defaultFailure()))
-      }
+  def makeOptOutUpdateRequest()(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[OptOutUpdateResponse] = {
+    fetchOptOutProposition().flatMap { proposition =>
+      proposition.optOutPropositionType.map {
+        case _: OneYearOptOutProposition =>
+          saveIntent(intent = proposition.availableTaxYearsForOptOut.head)
+          makeOptOutUpdateRequest(proposition)
+        case _: MultiYearOptOutProposition => makeOptOutUpdateRequest(proposition)
+      } getOrElse Future.successful(OptOutUpdateResponseFailure.defaultFailure())
     }
   }
 
+  def saveIntent(intent: TaxYear)(implicit hc: HeaderCarrier): Future[Boolean] = {
+    val data = UIJourneySessionData(
+      sessionId = hc.sessionId.get.value,
+      journeyType = OptOutJourney.Name,
+      optOutSessionData = Some(OptOutSessionData(selectedOptOutYear = Some(intent.toString)))
+    )
+    repository.set(data)
+  }
 
-  def makeOptOutUpdateRequest(optOutProposition: OptOutProposition, intent: OptOutTaxYear)
+  def makeOptOutUpdateRequest(optOutProposition: OptOutProposition)
                              (implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[OptOutUpdateResponse] = {
 
-    val optOutYearsToUpdate = optOutProposition.optOutYearsToUpdate(intent)
+    def makeUpdateCalls(optOutYearsToUpdate: Seq[OptOutTaxYear]): Seq[Future[OptOutUpdateResponse]] = {
+      optOutYearsToUpdate.map(optOutYear => itsaStatusUpdateConnector.requestOptOutForTaxYear(optOutYear.taxYear, user.nino, optOutUpdateReason))
+    }
 
-    val responses: Seq[Future[OptOutUpdateResponse]] = optOutYearsToUpdate.map(optOutYear =>
-      itsaStatusUpdateConnector.requestOptOutForTaxYear(optOutYear.taxYear, user.nino, optOutUpdateReason))
+    def findAnyFailOrFirstSuccess(responses: Seq[OptOutUpdateResponse]): OptOutUpdateResponse = {
+      responses.find(_.isInstanceOf[OptOutUpdateResponseFailure]).getOrElse(responses.head)
```

Comment:
us of isInstance or asInstance is discourage in general and used as a last option; you should be able to pattern match or use collect etc over the result, depending what you are trying to achieve

Reply:
discussed this with Siddharth and we can't think of a better way to do this, this is a very specific operation of finding any fail or the first success as the method name suggests

Reply:
you can do something like, also I am not sure if it is guarantied to have head item in the response
```
responses
        .find{
          case _: OptOutUpdateResponseFailure => true
          case _ => false }
        .getOrElse(responses.head)
```